File: library/kotlin/io/envoyproxy/envoymobile/stats/TimerImpl.kt
package io.envoyproxy.envoymobile

import io.envoyproxy.envoymobile.engine.EnvoyEngine
import java.lang.ref.WeakReference

/**
 * Envoy implementation of a `Timer` for time measurements e.g. distribution of durations.
 */
internal class TimerImpl : Timer {
  var envoyEngine: WeakReference<EnvoyEngine>
  var series: String
  var tags: Map<String, String>

  internal constructor(engine: EnvoyEngine, elements: List<Element>, tags: List<Tag>) {
    this.envoyEngine = WeakReference<EnvoyEngine>(engine)
    this.series = elements.joinToString(separator = ".") { it.value }
    this.tags = convert(tags)
  }

  private constructor(engine: WeakReference<EnvoyEngine>, series: String, tags: List<Tag>) {
    this.envoyEngine = engine
    this.series = series
    this.tags = convert(tags)
  }

  override fun completeWithDuration(durationMs: Int) {
    envoyEngine.get()?.recordHistogramDuration(series, tags, durationMs)
  }

  override fun attach(tags: List<Tag>): Timer {
    return TimerImpl(envoyEngine, series, tags)
  }
}

File: library/swift/stats/TimerImpl.swift
@_implementationOnly import EnvoyEngine
import Foundation

/// The implementation of timer that can be used to track a distribution of time durations
@objcMembers
final class TimerImpl: NSObject, Timer {
  private let series: String
  private let tags: [String: String]
  private weak var engine: EnvoyEngine?

  init(elements: [Element], tags: [Tag], engine: EnvoyEngine) {
    self.series = elements.map { $0.value }.joined(separator: ".")
    self.engine = engine
    self.tags = Stats.convert(tags: tags)
    super.init()
  }

  init(series: String, tags: [Tag], engine: EnvoyEngine?) {
    self.series = series
    self.engine = engine
    self.tags = Stats.convert(tags: tags)
    super.init()
  }

  /// Record a new duration value for the distribution.
  func completeWithDuration(durationMs: Int) {
    // TODO(jingwei99) potentially surface error up if engine is nil.
    self.engine?.recordHistogramDuration(
      self.series, tags: tags, durationMs: numericCast(durationMs))
  }

  func attach(tags: [Tag]) -> Timer {
    return TimerImpl(series: self.series, tags: tags, engine: self.engine)
  }
}
